extension/api.js

```diff
@@ -0,0 +1,154 @@
+/* global ExtensionCommon, ExtensionAPI, Services, XPCOMUtils */
+const { SearchUtils } = ChromeUtils.import(
+  "resource://gre/modules/SearchUtils.jsm"
+);
+const { WebRequest } = ChromeUtils.import(
+  "resource://gre/modules/WebRequest.jsm"
+);
+const { AddonManager } = ChromeUtils.import(
+  "resource://gre/modules/AddonManager.jsm"
+);
+
+XPCOMUtils.defineLazyGlobalGetters(this, ["ChannelWrapper"]);
+
+XPCOMUtils.defineLazyServiceGetter(
+  Services,
+  "eTLD",
+  "@mozilla.org/network/effective-tld-service;1",
+  "nsIEffectiveTLDService"
+);
+
+this.addonsSearchExperiment = class extends ExtensionAPI {
+  onStartup() {}
+
+  makeOnSearchEngineModified(context) {
+    return new ExtensionCommon.EventManager({
+      context,
+      name: "addonsSearchExperiment.onSearchEngineModified",
+      register: (fire) => {
+        const onSearchEngineModifiedObserver = {
+          observe(aSubject, aTopic, aData) {
+            if (aTopic !== SearchUtils.TOPIC_ENGINE_MODIFIED) {
+              return;
+            }
+
+            fire.async(aData);
+          },
+        };
+
+        Services.obs.addObserver(
```

Comment:
Might be good to wait for search initialization before adding the observer.  On firefox startup, that will prevent churn when the service is adding engines.

https://searchfox.org/mozilla-central/rev/9c9f2f85d00aef1943cb521ac95c72bae932ebc5/browser/components/extensions/parent/ext-browser.js#289